Description: Upstream changes introduced in version 0.1.0-0+oneiric1024~3beol
 This patch has been created by dpkg-source during the package build.
 Here's the last changelog entry, hopefully it gives details on why
 those changes were made:
 .
 libhangul (0.1.0-0+oneiric1024~3beol) oneiric; urgency=low
 .
   [floor]
   * 3beol
 .
 The person named in the Author field signed this changelog entry.
Author: floor <createsc@gmail.com>

---
The information above should follow the Patch Tagging Guidelines, please
checkout http://dep.debian.net/deps/dep3/ to learn about the format. Here
are templates for supplementary fields that you might want to add:

Origin: <vendor|upstream|other>, <url of original patch>
Bug: <url in upstream bugtracker>
Bug-Debian: http://bugs.debian.org/<bugnumber>
Bug-Ubuntu: https://launchpad.net/bugs/<bugnumber>
Forwarded: <no|not-needed|url proving that it has been forwarded>
Reviewed-By: <name and email of someone who approved the patch>
Last-Update: <YYYY-MM-DD>

--- libhangul-0.1.0.orig/hangul/hangul.h
+++ libhangul-0.1.0/hangul/hangul.h
@@ -87,6 +87,7 @@ enum {
 enum {
     HANGUL_KEYBOARD_TYPE_JAMO,
     HANGUL_KEYBOARD_TYPE_JASO,
+    HANGUL_KEYBOARD_TYPE_SHIN_JASO,
     HANGUL_KEYBOARD_TYPE_ROMAJA
 };
 
--- libhangul-0.1.0.orig/hangul/hangulinputcontext.c
+++ libhangul-0.1.0/hangul/hangulinputcontext.c
@@ -210,6 +210,8 @@ struct _HangulBuffer {
     ucschar jungseong;
     ucschar jongseong;
 
+    bool double_moeum;
+
     ucschar stack[12];
     int     index;
 };
@@ -242,6 +244,11 @@ static const HangulCombination hangul_co
     (HangulCombinationItem*)hangul_combination_table_default
 };
 
+static const HangulCombination hangul_combination_3ahnmatae = {
+    N_ELEMENTS(hangul_combination_table_3ahnmatae),
+    (HangulCombinationItem*)hangul_combination_table_3ahnmatae
+};
+
 static const HangulCombination hangul_combination_romaja = {
     N_ELEMENTS(hangul_combination_table_romaja),
     (HangulCombinationItem*)hangul_combination_table_romaja
@@ -297,6 +304,38 @@ static const HangulKeyboard hangul_keybo
     &hangul_combination_default
 };
 
+static const HangulKeyboard hangul_keyboard_3shin = {
+    HANGUL_KEYBOARD_TYPE_SHIN_JASO,
+    "3shin",
+    N_("Sebeolsik Shin"),
+    (ucschar*)hangul_keyboard_table_3shin,
+    &hangul_combination_default
+};
+
+static const HangulKeyboard hangul_keyboard_3ahnmatae = {
+    HANGUL_KEYBOARD_TYPE_JASO,
+    "3ahn",
+    N_("Sebeolsik 3Ahnmatae"),
+    (ucschar*)hangul_keyboard_table_3ahnmatae,
+    &hangul_combination_3ahnmatae
+};
+
+static const HangulKeyboard hangul_keyboard_2dvorak = {
+    HANGUL_KEYBOARD_TYPE_JAMO,
+    "2d",
+    N_("Dubeolsik Dvorak"),
+    (ucschar*)hangul_keyboard_table_2dvorak,
+    &hangul_combination_default
+};
+
+static const HangulKeyboard hangul_keyboard_3finaldvorak = {
+    HANGUL_KEYBOARD_TYPE_JASO,
+    "3fd",
+    N_("Sebeolsik Final Dvorak"),
+    (ucschar*)hangul_keyboard_table_3finaldvorak,
+    &hangul_combination_default
+};
+
 static const HangulKeyboard hangul_keyboard_3sun = {
     HANGUL_KEYBOARD_TYPE_JASO,
     "3s",
@@ -334,6 +373,10 @@ static const HangulKeyboard* hangul_keyb
     &hangul_keyboard_2y,
     &hangul_keyboard_390,
     &hangul_keyboard_3final,
+    &hangul_keyboard_3shin,
+    &hangul_keyboard_3ahnmatae,
+    &hangul_keyboard_2dvorak,
+    &hangul_keyboard_3finaldvorak,
     &hangul_keyboard_3sun,
     &hangul_keyboard_3yet,
     &hangul_keyboard_32,
@@ -571,6 +614,8 @@ hangul_buffer_clear(HangulBuffer *buffer
     buffer->jungseong = 0;
     buffer->jongseong = 0;
 
+    buffer->double_moeum = false;
+
     buffer->index = -1;
     buffer->stack[0]  = 0;
     buffer->stack[1]  = 0;
@@ -723,6 +768,9 @@ hangul_buffer_backspace(HangulBuffer *bu
 	    buffer->choseong = 0;
 	    buffer->jungseong = 0;
 	    buffer->jongseong = 0;
+
+            buffer->double_moeum = false;
+
 	    return true;
 	}
     }
@@ -1077,6 +1125,541 @@ hangul_ic_process_jaso(HangulInputContex
 }
 
 static bool
+hangul_ic_process_shin_jaso(HangulInputContext *hic, int ascii, ucschar ch)
+{
+    if (hangul_is_choseong(ch)) {
+        if(hic->buffer.choseong){//첫소리 있고, 첫소리가 들어왔다.
+            if(ch == 0x1106 || ch == 0x110e || ch == 0x1111){//1106(ㅁ) 110e(ㅊ) 1111(ㅍ) //첫소리 있고, ㅁ, ㅊ, ㅍ 가 들어왔다.
+                if(hic->buffer.jungseong){//첫소리, 가윗소리가 있다. 들어온 첫소리를 첫소리로 한다.
+                        ucschar choseong = 0;
+                        if (hangul_is_choseong(hangul_ic_peek(hic))) {
+                            choseong = hangul_combination_combine(hic->keyboard->combination,
+                                hic->buffer.choseong, ch);
+                        }
+                        if (choseong) {
+                            if (!hangul_ic_push(hic, choseong)) {
+                                if (!hangul_ic_push(hic, choseong)) {
+                                    return false;
+                                }
+                            }
+                        } else {
+                            hangul_ic_save_commit_string(hic);
+                            if (!hangul_ic_push(hic, ch)) {
+                                return false;
+                            }
+                        }
+                }else{//첫소리 있고, 가윗소리가 비었다. 들어온 첫소리를 가윗소리로 바꾼다.
+                    ch = hangul_shin_choseong_to_shin_jungseong(ch);
+                    if(ch == 0x1169 || ch == 0x116e){//1169(ㅗ) 116e(ㅜ)
+                        hic->buffer.double_moeum = true;
+                    }
+                    if (!hangul_ic_push(hic, ch)) {
+                        if (!hangul_ic_push(hic, ch)) {
+                            return false;
+                        }
+                    }
+                }
+            }else{//첫소리 있고, ㅁ, ㅊ, ㅍ 가 아닌 첫소리가 들어와서 첫소리로 다룬다.
+                    ucschar choseong = 0;
+                    if (hangul_is_choseong(hangul_ic_peek(hic))) {
+                        choseong = hangul_combination_combine(hic->keyboard->combination,
+                            hic->buffer.choseong, ch);
+                    }
+                    if (choseong) {
+                        if (!hangul_ic_push(hic, choseong)) {
+                            if (!hangul_ic_push(hic, choseong)) {
+                                return false;
+                            }
+                        }
+                    } else {
+                        hangul_ic_save_commit_string(hic);
+                        if (!hangul_ic_push(hic, ch)) {
+                            return false;
+                        }
+                    }
+            }
+        }else{//첫소리 없고, 들어온 첫소리를 첫소리로 다룬다.
+                if (!hangul_ic_push(hic, ch)) {
+                    if (!hangul_ic_push(hic, ch)) {
+                        return false;
+                    }
+                }
+        }
+    } else if (hangul_is_jungseong(ch)) {
+        if (hic->buffer.jungseong == 0) {//가윗소리가 없다. 들어온 가윗소리를 가윗소리로 다룬다.
+            if(ascii == 'O' || ascii == 'P'){//들어온 겹모음용 ㅜ, ㅗ 가윗 소리를 겹모음용으로 다룬다.//Shift + key
+                        hic->buffer.double_moeum = true;
+            }
+            if (!hangul_ic_push(hic, ch)) {
+                if (!hangul_ic_push(hic, ch)) {
+                    return false;
+                }
+            }
+        } else {//가윗소리가 있다. 가윗소리가 들어왔다.
+            if(hic->buffer.double_moeum){//겹모음용 가윗소리가 있다,
+                if(ascii == 'O' || ascii == 'P'){//들어온 겹모음용 ㅜ, ㅗ 가윗 소리를 겹모음용으로 다룬다.
+                    ucschar jungseong = 0;
+
+                    if (hangul_is_jungseong(hangul_ic_peek(hic))) {
+                        jungseong = hangul_combination_combine(hic->keyboard->combination,
+                            hic->buffer.jungseong, ch);
+                    }
+                    if (jungseong) {
+                        if (!hangul_ic_push(hic, jungseong)) {
+                            if (!hangul_ic_push(hic, jungseong)) {
+                                        hic->buffer.double_moeum = false;
+                                return false;
+                            }
+                        }
+                    } else {
+                        hangul_ic_save_commit_string(hic);
+                        if (!hangul_ic_push(hic, ch)) {
+                            if (!hangul_ic_push(hic, ch)) {
+                                        hic->buffer.double_moeum = true;
+                                return false;
+                            }
+                        }
+                    }
+                }else{//겹모음용 가윗소리가 있고, 홑모음용 가윗소리가 들어왔다.
+                    ucschar jungseong = 0;
+
+                    if (hangul_is_jungseong(hangul_ic_peek(hic))) {
+                        jungseong = hangul_combination_combine(hic->keyboard->combination,
+                            hic->buffer.jungseong, ch);
+                    }
+                    if (jungseong) {
+                        if (!hangul_ic_push(hic, jungseong)) {
+                            if (!hangul_ic_push(hic, jungseong)) {
+                                        hic->buffer.double_moeum = false;
+                                return false;
+                            }
+                        }
+                    } else {
+                        hangul_ic_save_commit_string(hic);
+                        if (!hangul_ic_push(hic, ch)) {
+                            if (!hangul_ic_push(hic, ch)) {
+                                            hic->buffer.double_moeum = false;
+                                return false;
+                            }
+                        }
+                    }
+                }
+            }else{//홑모음 가윗소리가 있다.
+                if(ascii == 'O' || ascii == 'P'){//들어온 겹모음용 ㅜ, ㅗ 가윗 소리를 겹모음용으로 다룬다.
+                    ucschar jungseong = 0;
+
+                    if (hangul_is_jungseong(hangul_ic_peek(hic))) {
+                        jungseong = hangul_combination_combine(hic->keyboard->combination,
+                            hic->buffer.jungseong, ch);
+                    }
+                    if (jungseong) {//홑모음용 가윗소리가 있다. 겹모음용 ㅜ, ㅗ 가윗소리가 들어왔다.
+                        if (!hangul_ic_push(hic, jungseong)) {
+                            if (!hangul_ic_push(hic, jungseong)) {
+                                        hic->buffer.double_moeum = false;
+                                return false;
+                            }
+                        }
+                    } else {//가윗소리 홑모음용이 있고, 겹모음용 가윗소리가 들어왔지만 모으지 못했다.
+                        hangul_ic_save_commit_string(hic);
+                        if (!hangul_ic_push(hic, ch)) {
+                            if (!hangul_ic_push(hic, ch)) {
+                                    hic->buffer.double_moeum = true;
+                                return false;
+                            }
+                        }
+                    }
+                }else{//홑모음 가윗소리가 있다. 홑모음 가윗소리가 들어왔다.
+                    ucschar jungseong = 0;
+
+                    if (hangul_is_jungseong(hangul_ic_peek(hic))) {
+                        jungseong = hangul_combination_combine(hic->keyboard->combination,
+                            hic->buffer.jungseong, ch);
+                    }
+                    if (jungseong) {
+                        if (!hangul_ic_push(hic, jungseong)) {
+                            if (!hangul_ic_push(hic, jungseong)) {
+                                    hic->buffer.double_moeum = false;
+                                return false;
+                            }
+                        }
+                    } else {//가윗소리 홑모음용이 있고, 가윗소리가 들어왔다.
+                        hangul_ic_save_commit_string(hic);
+                        if (!hangul_ic_push(hic, ch)) {
+                            if (!hangul_ic_push(hic, ch)) {
+                                    hic->buffer.double_moeum = false;
+                                return false;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    } else if (hangul_is_jongseong(ch)) {
+        if(hic->buffer.choseong){//첫소리가 있다.
+            if(hic->buffer.jungseong){//첫소리, 가윗소리가 있다.
+//                if(hic->buffer.jungseong == 0x1169 || hic->buffer.jungseong == 0x116e){//1169(ㅗ) 116e(ㅜ) //첫소리, 가윗소리 ㅜ, ㅗ 가 있다. 끝소리가 들어왔다.
+                if(hic->buffer.double_moeum){//1169(ㅗ) 116e(ㅜ) //가윗소리에 겹모음용이 들어있다.
+                    if(ch == 0x11b7){//끝소리가 ᆷ 이면 끝소리로 다룬다.
+                        if (hic->buffer.jongseong == 0) {//첫소리, 가윗소리가 있다. 끝소리를 넣는다.
+                            if (!hangul_ic_push(hic, ch)) {
+                                if (!hangul_ic_push(hic, ch)) {
+                                    return false;
+                                }
+                            }
+                        } else {//첫소리, 가윗소리, 끝소리가 있다.
+                            ucschar jongseong = 0;
+                            if (hangul_is_jongseong(hangul_ic_peek(hic))) {
+                                jongseong = hangul_combination_combine(hic->keyboard->combination,
+                                    hic->buffer.jongseong, ch);
+                            }
+                            if (jongseong) {
+                                if (!hangul_ic_push(hic, jongseong)) {
+                                    if (!hangul_ic_push(hic, jongseong)) {
+                                        return false;
+                                    }
+                                }
+                            } else {
+                                hangul_ic_save_commit_string(hic);
+                                if (!hangul_ic_push(hic, ch)) {
+                                    if (!hangul_ic_push(hic, ch)) {
+                                            hic->buffer.double_moeum = false;
+                                        return false;
+                                    }
+                                }
+                            }
+                        }
+                    }else{//첫소리, 겹모음 가윗소리가 있다. 들어온 끝소리를 홑모음 가윗소리로 다룬다.
+                        ucschar new_temp = ch;
+                        ch = hangul_shin_jongseong_to_shin_jungseong(ch);
+
+                        ucschar jungseong = 0;
+                        if (hangul_is_jungseong(hangul_ic_peek(hic))) {
+                            jungseong = hangul_combination_combine(hic->keyboard->combination,
+                                hic->buffer.jungseong, ch);
+                        }
+                        if (jungseong) {//첫소리, 가윗소리 ㅜ, ㅗ 가 겹모음용이다. 끝소리를 가윗소리로 바꾸어 합친다.
+                            if (!hangul_ic_push(hic, jungseong)) {
+                                if (!hangul_ic_push(hic, jungseong)) {
+                                        hic->buffer.double_moeum = false;
+                                    return false;
+                                }
+                            }
+                        } else {//첫소리가 있고, 겹모음 가윗소리가 있지만, 모음 합치기가 안되어 끝소리로 다룬다.
+                            ch = new_temp;
+
+                            if (hic->buffer.jongseong == 0) {//첫소리, 가윗소리가 있다. 끝소리를 넣는다.
+                                if (!hangul_ic_push(hic, ch)) {
+                                    if (!hangul_ic_push(hic, ch)) {
+                                        return false;
+                                    }
+                                }
+                            } else {//첫소리, 가윗소리, 끝소리가 있다.
+                                ucschar jongseong = 0;
+                                if (hangul_is_jongseong(hangul_ic_peek(hic))) {
+                                    jongseong = hangul_combination_combine(hic->keyboard->combination,
+                                        hic->buffer.jongseong, ch);
+                                }
+                                if (jongseong) {
+                                    if (!hangul_ic_push(hic, jongseong)) {
+                                        if (!hangul_ic_push(hic, jongseong)) {
+                                            return false;
+                                        }
+                                    }
+                                } else {
+                                    hangul_ic_save_commit_string(hic);
+                                    if (!hangul_ic_push(hic, ch)) {
+                                        if (!hangul_ic_push(hic, ch)) {
+                                                hic->buffer.double_moeum = false;
+                                            return false;
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }else{//첫소리, 홑모음용 가윗소리가 있다. 들어온 끝소리를 끝소리로 다룬다.
+                    if (hic->buffer.jongseong == 0) {//끝소리 처리 루틴;
+                        if (!hangul_ic_push(hic, ch)) {
+                            if (!hangul_ic_push(hic, ch)) {
+                                return false;
+                            }
+                        }
+                    } else {
+                        ucschar jongseong = 0;
+                        if (hangul_is_jongseong(hangul_ic_peek(hic))) {
+                            jongseong = hangul_combination_combine(hic->keyboard->combination,
+                                hic->buffer.jongseong, ch);
+                        }
+                        if (jongseong) {
+                            if (!hangul_ic_push(hic, jongseong)) {
+                                if (!hangul_ic_push(hic, jongseong)) {
+                                    return false;
+                                }
+                            }
+                        } else {
+                            hangul_ic_save_commit_string(hic);
+                            if (!hangul_ic_push(hic, ch)) {
+                                if (!hangul_ic_push(hic, ch)) {
+                                    return false;
+                                }
+                            }
+                        }
+                    }
+                }
+            }else{//첫소리가 있고, 가윗소리가 없다.
+                if(ch == 0x11b7){//11b7 (ᆷ) // 첫소리가 있고, 가윗소리가 없다. 끝소리 ᆷ 이 들어왔다.
+                    if (hic->buffer.jongseong == 0) {//끝소리 처리 루틴;
+                        if (!hangul_ic_push(hic, ch)) {
+                            if (!hangul_ic_push(hic, ch)) {
+                                return false;
+                            }
+                        }
+                    } else {
+                        ucschar jongseong = 0;
+                        if (hangul_is_jongseong(hangul_ic_peek(hic))) {
+                            jongseong = hangul_combination_combine(hic->keyboard->combination,
+                                    hic->buffer.jongseong, ch);
+                        }
+                        if (jongseong) {
+                            if (!hangul_ic_push(hic, jongseong)) {
+                                if (!hangul_ic_push(hic, jongseong)) {
+                                    return false;
+                                }
+                            }
+                        } else {
+                            hangul_ic_save_commit_string(hic);
+                            if (!hangul_ic_push(hic, ch)) {
+                                if (!hangul_ic_push(hic, ch)) {
+                                    return false;
+                                }
+                            }
+                        }
+                    }
+                }else{//첫소리가 있고, 가윗소리가 없다. 들어온 끝소리를 가윗소리로 다룬다.
+                        ch = hangul_shin_jongseong_to_shin_jungseong(ch);
+
+                        if (!hangul_ic_push(hic, ch)) {
+                            if (!hangul_ic_push(hic, ch)) {
+                                    hic->buffer.double_moeum = false;
+                                return false;
+                            }
+                        }
+                }
+            }
+        }else{//첫소리가 없다. 끝소리가 들어왔다.
+            if(hic->buffer.jungseong){//첫소리가 없고, 가윗소리가 있다.
+//                if(hic->buffer.jungseong == 0x1169 || hic->buffer.jungseong == 0x116e){//1169(ㅗ) 116e(ㅜ) //첫소리, 가윗소리 ㅜ, ㅗ 가 있다. 끝소리가 들어왔다.
+                if(hic->buffer.double_moeum){//1169(ㅗ) 116e(ㅜ) //가윗소리에 겹모음용이 들어있다.
+                    if(ch == 0x11b7){//끝소리가 ᆷ 이면 끝소리로 다룬다.
+                        if (hic->buffer.jongseong == 0) {//첫소리 없고, 가윗소리가 있다. 끝소리를 넣는다.
+                            if (!hangul_ic_push(hic, ch)) {
+                                if (!hangul_ic_push(hic, ch)) {
+                                    return false;
+                                }
+                            }
+                        } else {//첫소리, 가윗소리, 끝소리가 있다.
+                            ucschar jongseong = 0;
+                            if (hangul_is_jongseong(hangul_ic_peek(hic))) {
+                                jongseong = hangul_combination_combine(hic->keyboard->combination,
+                                    hic->buffer.jongseong, ch);
+                            }
+                            if (jongseong) {
+                                if (!hangul_ic_push(hic, jongseong)) {
+                                    if (!hangul_ic_push(hic, jongseong)) {
+                                        return false;
+                                    }
+                                }
+                            } else {
+                                hangul_ic_save_commit_string(hic);
+                                if (!hangul_ic_push(hic, ch)) {
+                                    if (!hangul_ic_push(hic, ch)) {
+                                            hic->buffer.double_moeum = false;
+                                        return false;
+                                    }
+                                }
+                            }
+                        }
+                    }else{//첫소리없고, 겹모음 가윗소리가 있다. 들어온 끝소리를 홑모음 가윗소리로 다룬다.
+                        ucschar new_temp = ch;
+                        ch = hangul_shin_jongseong_to_shin_jungseong(ch);
+
+                        ucschar jungseong = 0;
+                        if (hangul_is_jungseong(hangul_ic_peek(hic))) {
+                            jungseong = hangul_combination_combine(hic->keyboard->combination,
+                                hic->buffer.jungseong, ch);
+                        }
+                        if (jungseong) {//첫소리없고, 가윗소리 ㅜ, ㅗ 가 겹모음용이다. 끝소리를 가윗소리로 바꾸어 합친다.
+                            if (!hangul_ic_push(hic, jungseong)) {
+                                if (!hangul_ic_push(hic, jungseong)) {
+                                        hic->buffer.double_moeum = false;
+                                    return false;
+                                }
+                            }
+                        } else {//첫소리가 없고, 겹모음 가윗소리가 있지만, 모음 합치기가 안되어 끝소리로 다룬다.
+                            ch = new_temp;
+
+                            if (hic->buffer.jongseong == 0) {//첫소리 없고, 가윗소리가 있다. 끝소리를 넣는다.
+                                if (!hangul_ic_push(hic, ch)) {
+                                    if (!hangul_ic_push(hic, ch)) {
+                                        return false;
+                                    }
+                                }
+                            } else {//첫소리 없고, 가윗소리, 끝소리가 있다.
+                                ucschar jongseong = 0;
+                                if (hangul_is_jongseong(hangul_ic_peek(hic))) {
+                                    jongseong = hangul_combination_combine(hic->keyboard->combination,
+                                        hic->buffer.jongseong, ch);
+                                }
+                                if (jongseong) {
+                                    if (!hangul_ic_push(hic, jongseong)) {
+                                        if (!hangul_ic_push(hic, jongseong)) {
+                                            return false;
+                                        }
+                                    }
+                                } else {
+                                    hangul_ic_save_commit_string(hic);
+                                    if (!hangul_ic_push(hic, ch)) {
+                                        if (!hangul_ic_push(hic, ch)) {
+                                                hic->buffer.double_moeum = false;
+                                            return false;
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }else{//첫소리 없고, 홑모음용 가윗소리가 있다. 들어온 끝소리를 끝소리로 다룬다.
+                    if (hic->buffer.jongseong == 0) {//끝소리 처리 루틴;
+                        if (!hangul_ic_push(hic, ch)) {
+                            if (!hangul_ic_push(hic, ch)) {
+                                return false;
+                            }
+                        }
+                    } else {
+                        ucschar jongseong = 0;
+                        if (hangul_is_jongseong(hangul_ic_peek(hic))) {
+                            jongseong = hangul_combination_combine(hic->keyboard->combination,
+                                hic->buffer.jongseong, ch);
+                        }
+                        if (jongseong) {
+                            if (!hangul_ic_push(hic, jongseong)) {
+                                if (!hangul_ic_push(hic, jongseong)) {
+                                    return false;
+                                }
+                            }
+                        } else {
+                            hangul_ic_save_commit_string(hic);
+                            if (!hangul_ic_push(hic, ch)) {
+                                if (!hangul_ic_push(hic, ch)) {
+                                    return false;
+                                }
+                            }
+                        }
+                    }
+                }
+            }else{//첫소리가 없고, 가윗소리가 없다. 끝소리가 들어왔다.
+                if (hic->buffer.jongseong == 0) {//끝소리 처리 루틴;
+                    if (!hangul_ic_push(hic, ch)) {
+                        if (!hangul_ic_push(hic, ch)) {
+                            return false;
+                        }
+                    }
+                } else {
+                    ucschar jongseong = 0;
+                    if (hangul_is_jongseong(hangul_ic_peek(hic))) {
+                        jongseong = hangul_combination_combine(hic->keyboard->combination,
+						   hic->buffer.jongseong, ch);
+                    }
+                    if (jongseong) {
+                        if (!hangul_ic_push(hic, jongseong)) {
+                            if (!hangul_ic_push(hic, jongseong)) {
+                                return false;
+                            }
+                        }
+                    } else {
+                        hangul_ic_save_commit_string(hic);
+                        if (!hangul_ic_push(hic, ch)) {
+                            if (!hangul_ic_push(hic, ch)) {
+                                return false;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    } else if (ch > 0) {
+        hangul_ic_save_commit_string(hic);
+        hic->buffer.double_moeum = false;
+        hangul_ic_append_commit_string(hic, ch);
+    } else {
+        hangul_ic_save_commit_string(hic);
+        hic->buffer.double_moeum = false;
+        return false;
+    }
+
+    hangul_ic_save_preedit_string(hic);
+    return true;
+}
+
+/*
+	if (hic->buffer.choseong == 0) {//첫소리 처리 루틴;
+	    if (!hangul_ic_push(hic, ch)) {
+            if (!hangul_ic_push(hic, ch)) {
+                return false;
+            }
+	    }
+	} else {
+	    ucschar choseong = 0;
+	    if (hangul_is_choseong(hangul_ic_peek(hic))) {
+            choseong = hangul_combination_combine(hic->keyboard->combination,
+						  hic->buffer.choseong, ch);
+	    }
+	    if (choseong) {
+            if (!hangul_ic_push(hic, choseong)) {
+                if (!hangul_ic_push(hic, choseong)) {
+                    return false;
+                }
+            }
+	    } else {
+            hangul_ic_save_commit_string(hic);
+            if (!hangul_ic_push(hic, ch)) {
+                return false;
+            }
+	    }
+	}
+	*/
+
+/*
+        if (hic->buffer.jongseong == 0) {//끝소리 처리 루틴;
+            if (!hangul_ic_push(hic, ch)) {
+                if (!hangul_ic_push(hic, ch)) {
+                    return false;
+                }
+            }
+        } else {
+            ucschar jongseong = 0;
+            if (hangul_is_jongseong(hangul_ic_peek(hic))) {
+                jongseong = hangul_combination_combine(hic->keyboard->combination,
+						   hic->buffer.jongseong, ch);
+            }
+            if (jongseong) {
+                if (!hangul_ic_push(hic, jongseong)) {
+                    if (!hangul_ic_push(hic, jongseong)) {
+                        return false;
+                    }
+                }
+            } else {
+                hangul_ic_save_commit_string(hic);
+                if (!hangul_ic_push(hic, ch)) {
+                    if (!hangul_ic_push(hic, ch)) {
+                        return false;
+                    }
+                }
+            }
+        }
+*/
+
+static bool
 hangul_ic_process_romaja(HangulInputContext *hic, int ascii, ucschar ch)
 {
     ucschar jong;
@@ -1299,6 +1882,8 @@ hangul_ic_process(HangulInputContext *hi
 	return hangul_ic_process_jamo(hic, c);
     else if (hangul_keyboard_get_type(hic->keyboard) == HANGUL_KEYBOARD_TYPE_JASO)
 	return hangul_ic_process_jaso(hic, c);
+    else if (hangul_keyboard_get_type(hic->keyboard) == HANGUL_KEYBOARD_TYPE_SHIN_JASO)
+        return hangul_ic_process_shin_jaso(hic, ascii, c);
     else
 	return hangul_ic_process_romaja(hic, ascii, c);
 }
--- libhangul-0.1.0.orig/hangul/hangulkeyboard.h
+++ libhangul-0.1.0/hangul/hangulkeyboard.h
@@ -1,3 +1,786 @@
+static const ucschar hangul_keyboard_table_3shin[] = {
+    0x0000    ,        /*    0x00    :    null    :        */
+    0x0000    ,        /*    0x01    :    start of heading    :        */
+    0x0000    ,        /*    0x02    :    start of text    :        */
+    0x0000    ,        /*    0x03    :    end of text    :        */
+    0x0000    ,        /*    0x04    :    end of transmission    :        */
+    0x0000    ,        /*    0x05    :    enquiry    :        */
+    0x0000    ,        /*    0x06    :    acknowledge    :        */
+    0x0000    ,        /*    0x07    :    bell    :        */
+    0x0000    ,        /*    0x08    :    backspace    :        */
+    0x0000    ,        /*    0x09    :    character tabulation    :        */
+    0x0000    ,        /*    0x0A    :    line feed (lf)    :        */
+    0x0000    ,        /*    0x0B    :    line tabulation    :        */
+    0x0000    ,        /*    0x0C    :    form feed (ff)    :        */
+    0x0000    ,        /*    0x0D    :    carriage return (cr)    :        */
+    0x0000    ,        /*    0x0E    :    shift out    :        */
+    0x0000    ,        /*    0x0F    :    shift in    :        */
+    0x0000    ,        /*    0x10    :    data link escape    :        */
+    0x0000    ,        /*    0x11    :    device control one    :        */
+    0x0000    ,        /*    0x12    :    device control two    :        */
+    0x0000    ,        /*    0x13    :    device control three    :        */
+    0x0000    ,        /*    0x14    :    device control four    :        */
+    0x0000    ,        /*    0x15    :    negative acknowledge    :        */
+    0x0000    ,        /*    0x16    :    synchronous idle    :        */
+    0x0000    ,        /*    0x17    :    end of transmission block    :        */
+    0x0000    ,        /*    0x18    :    cancel    :        */
+    0x0000    ,        /*    0x19    :    end of medium    :        */
+    0x0000    ,        /*    0x1A    :    substitute    :        */
+    0x0000    ,        /*    0x1B    :    escape    :        */
+    0x0000    ,        /*    0x1C    :    information separator four    :        */
+    0x0000    ,        /*    0x1D    :    information separator three    :        */
+    0x0000    ,        /*    0x1E    :    information separator two    :        */
+    0x0000    ,        /*    0x1F    :    information separator one    :        */
+    0x0000    ,        /*    0x20    :    space    :        */
+    0x0021    ,    /* ! */    /*    0x21    :    exclam    :    exclamation mark    */
+    0x0022    ,    /* " */    /*    0x22    :    quotedbl    :    quotation mark    */
+    0x0023    ,    /* # */    /*    0x23    :    numbersign    :    number sign    */
+    0x0024    ,    /* $ */    /*    0x24    :    dollar    :    dollar sign    */
+    0x0025    ,    /* % */    /*    0x25    :    percent    :    percent sign    */
+    0x0026    ,    /* & */    /*    0x26    :    ampersand    :    ampersand    */
+    0x1110    ,    /* ' */    /*    0x27    :    apostrophe    :    Choseong thieuth    */
+    0x0028    ,    /* ( */    /*    0x28    :    parenleft    :    left parenthesis    */
+    0x0029    ,    /* ) */    /*    0x29    :    parenright    :    right parenthesis    */
+    0x002A    ,    /* * */    /*    0x2A    :    asterisk    :    asterisk    */
+    0x002B    ,    /* + */    /*    0x2B    :    plus    :    plus sign    */
+    0x002C    ,    /* , */    /*    0x2C    :    comma    :    comma    */
+    0x002D    ,    /* - */    /*    0x2D    :    minus    :    minus sign    */
+    0x002E    ,    /* . */    /*    0x2E    :    period    :    period    */
+    0x110f    ,    /* / */    /*    0x2F    :    slash    :    choseong khieukh    */
+    0x0030    ,    /* 0 */    /*    0x30    :    0    :    0    */
+    0x0031    ,    /* 1 */    /*    0x31    :    1    :    1    */
+    0x0032    ,    /* 2 */    /*    0x32    :    2    :    2    */
+    0x0033    ,    /* 3 */    /*    0x33    :    3    :    3    */
+    0x0034    ,    /* 4 */    /*    0x34    :    4    :    4    */
+    0x0035    ,    /* 5 */    /*    0x35    :    5    :    5    */
+    0x0036    ,    /* 6 */    /*    0x36    :    6    :    6    */
+    0x0037    ,    /* 7 */    /*    0x37    :    7    :    7    */
+    0x0038    ,    /* 8 */    /*    0x38    :    8    :    8    */
+    0x0039    ,    /* 9 */    /*    0x39    :    9    :    9    */
+    0x003A    ,    /* : */    /*    0x3A    :    colon    :    colon    */
+    0x1107    ,    /* ; */    /*    0x3B    :    semicolon    :    Choseong pieup    */
+    0x003c    ,    /* < */    /*    0x3C    :    less    :    less-than sign    */
+    0x003d    ,    /* = */    /*    0x3D    :    equal    :    equals sign    */
+    0x003e    ,    /* > */    /*    0x3E    :    greater    :    greater-than sign    */
+    0x003f    ,    /* ? */    /*    0x3F    :    question    :    question mark    */
+    0x0040    ,    /* @ */    /*    0x40    :    at    :    commercial at    */
+    0x1172    ,    /* A */    /*    0x41    :    A    :    jungseong yu    */
+    0x116e    ,    /* B */    /*    0x42    :    B    :    jungseong u    */
+    0x1166    ,    /* C */    /*    0x43    :    C    :    jungseong e    */
+    0x1175    ,    /* D */    /*    0x44    :    D    :    jungseong I    */
+    0x1167    ,    /* E */    /*    0x45    :    E    :    jungseong yeo    */
+    0x1161    ,    /* F */    /*    0x46    :    F    :    jungseong a    */
+    0x1173    ,    /* G */    /*    0x47    :    G    :    jungseong eu    */
+    0x2018    ,    /* H */    /*    0x48    :    H    :    quota left    */
+    0x1174    ,    /* I */    /*    0x49    :    I    :    jungseong yeu    */
+    0x2019    ,    /* J */    /*    0x4A    :    J    :    quota right    */
+    0x003b    ,    /* K */    /*    0x4B    :    K    :    semicolon    */
+    0x0027    ,    /* L */    /*    0x4C    :    L    :    quota    */
+    0x002F    ,    /* M */    /*    0x4D    :    M    :    slash    */
+    0x02d0    ,    /* N */    /*    0x4E    :    N    :    ː 0x02d0 <- 0x119e jungseong arae a    */
+    0x116e    ,    /* O */    /*    0x4F    :    O    :    jungseong u    */
+    0x1169    ,    /* P */    /*    0x50    :    P    :    jungseong o    */
+    0x1164    ,    /* Q */    /*    0x51    :    Q    :    jungseong yae    */
+    0x1162    ,    /* R */    /*    0x52    :    R    :    jungseong yeo    */
+    0x1168    ,    /* S */    /*    0x53    :    S    :    jungseong ye    */
+    0x1165    ,    /* T */    /*    0x54    :    T    :    jungseong eo    */
+    0x201d    ,    /* U */    /*    0x55    :    U    :    quotadouble right    */
+    0x1169    ,    /* V */    /*    0x56    :    V    :    jungseong o    */
+    0x1163    ,    /* W */    /*    0x57    :    W    :    jungseong ya    */
+    0x116d    ,    /* X */    /*    0x58    :    X    :    jungseong yo    */
+    0x201c    ,    /* Y */    /*    0x59    :    Y    :    quotadouble left    */
+    0x203b    ,    /* Z */    /*    0x5A    :    Z    :    reference mark    */
+    0x005b    ,    /* [ */    /*    0x5B    :    bracketleft    :    left bracket    */
+    0x005c    ,    /* \ */    /*    0x5C    :    backslash    :    backslash    */
+    0x005d    ,    /* ] */    /*    0x5D    :    bracketright    :    right bracket    */
+    0x005e    ,    /* ^ */    /*    0x5E    :    asciicircum    :    circumflex accent    */
+    0x005f    ,    /* _ */    /*    0x5F    :    underscore    :    underscore    */
+    0x0060    ,    /* ` */    /*    0x60    :    quoteleft    :    grave accent    */
+    0x11bc    ,    /* a */    /*    0x61    :    a    :    Jongseong ieung    */
+    0x11bf    ,    /* b */    /*    0x62    :    b    :    Jongseong khiyeok    */
+    0x11be    ,    /* c */    /*    0x63    :    c    :    Jongseong chieuch    */
+    0x11ae    ,    /* d */    /*    0x64    :    d    :    Jongseong tikeut    */
+    0x11b8    ,    /* e */    /*    0x65    :    e    :    Jongseong pieup    */
+    0x11bb    ,    /* f */    /*    0x66    :    f    :    Jongseong ssangsios    */
+    0x11bd    ,    /* g */    /*    0x67    :    g    :    Jongseong cieuc    */
+    0x1102    ,    /* h */    /*    0x68    :    h    :    Choseong nieun    */
+    0x1106    ,    /* i */    /*    0x69    :    i    :    Choseong mieum    */
+    0x110b    ,    /* j */    /*    0x6A    :    j    :    Choseong ieung    */
+    0x1100    ,    /* k */    /*    0x6B    :    k    :    Choseong kiyeok    */
+    0x110c    ,    /* l */    /*    0x6C    :    l    :    Choseong cieuc    */
+    0x1112    ,    /* m */    /*    0x6D    :    m    :    Choseong hieuh    */
+    0x1109    ,    /* n */    /*    0x6E    :    n    :    Choseong sios    */
+    0x110e    ,    /* o */    /*    0x6F    :    o    :    Choseong chieuch    */
+    0x1111    ,    /* p */    /*    0x70    :    p    :    Choseong phieuph    */
+    0x11ba    ,    /* q */    /*    0x71    :    q    :    Jongseong sios    */
+    0x11c0    ,    /* r */    /*    0x72    :    r    :    Jongseong thikeuth    */
+    0x11ab    ,    /* s */    /*    0x73    :    s    :    Jongseong nieun    */
+    0x11c1    ,    /* t */    /*    0x74    :    t    :    Jongseong phieuph    */
+    0x1103    ,    /* u */    /*    0x75    :    u    :    Choseong tikeut    */
+    0x11c2    ,    /* v */    /*    0x76    :    v    :    Jongseong hieuh    */
+    0x11af    ,    /* w */    /*    0x77    :    w    :    Jongseong rieul    */
+    0x11a8    ,    /* x */    /*    0x78    :    x    :    Jongseong kiyeok    */
+    0x1105    ,    /* y */    /*    0x79    :    y    :    Choseong rieul    */
+    0x11b7    ,    /* z */    /*    0x7A    :    z    :    Jongseong mieum    */
+    0x007b    ,    /* { */    /*    0x7B    :    braceleft    :    left brace    */
+    0x007c    ,    /* | */    /*    0x7C    :    bar    :    vertical bar    */
+    0x007d    ,    /* } */    /*    0x7D    :    braceright    :    right brace    */
+    0x007e    ,    /* ~ */    /*    0x7E    :    asciitilde    :    tilde    */
+    0x0000            /*    0x7F    :    delete    :        */
+};
+
+static const ucschar hangul_keyboard_table_2dvorak[] = {
+    0x0000,    /* 0x00 null                                        */
+    0x0000,    /* 0x01 start of heading                            */
+    0x0000,    /* 0x02 start of text                                */
+    0x0000,    /* 0x03 end of text                                */
+    0x0000,    /* 0x04 end of transmission                        */
+    0x0000,    /* 0x05 enquiry                                    */
+    0x0000,    /* 0x06 acknowledge                                */
+    0x0000,    /* 0x07 bell                                        */
+    0x0000,    /* 0x08 backspace                                    */
+    0x0000,    /* 0x09 character tabulation                        */
+    0x0000,    /* 0x0A line feed (lf)                             */
+    0x0000,    /* 0x0B line tabulation                            */
+    0x0000,    /* 0x0C form feed (ff)                             */
+    0x0000,    /* 0x0D carriage return (cr)                        */
+    0x0000,    /* 0x0E shift out                                    */
+    0x0000,    /* 0x0F shift in                                    */
+    0x0000,    /* 0x10 data link escape                            */
+    0x0000,    /* 0x11 device control one                         */
+    0x0000,    /* 0x12 device control two                         */
+    0x0000,    /* 0x13 device control three                        */
+    0x0000,    /* 0x14 device control four                        */
+    0x0000,    /* 0x15 negative acknowledge                        */
+    0x0000,    /* 0x16 synchronous idle                            */
+    0x0000,    /* 0x17 end of transmission block                    */
+    0x0000,    /* 0x18 cancel                                     */
+    0x0000,    /* 0x19 end of medium                                */
+    0x0000,    /* 0x1A substitute                                 */
+    0x0000,    /* 0x1B escape                                     */
+    0x0000,    /* 0x1C information separator four                 */
+    0x0000,    /* 0x1D information separator three                */
+    0x0000,    /* 0x1E information separator two                    */
+    0x0000,    /* 0x1F information separator one                    */
+    0x0000,    /* 0x20 space                                        */
+    0x0021,    /* ! 0x21 exclam:     exclamation mark             */
+    0x1108,    /* " 0x22 quotedbl:    quotation mark                */
+    0x0023,    /* # 0x23 numbersign:   number sign                    */
+    0x0024,    /* $ 0x24 dollar:     dollar sign                    */
+    0x0025,    /* % 0x25 percent:    percent sign                 */
+    0x0026,    /* & 0x26 ampersand:    ampersand                    */
+    0x1107,    /* ' 0x27 apostrophe:   apostrophe                    */
+    0x0028,    /* ( 0x28 parenleft:    left parenthesis             */
+    0x0029,    /* ) 0x29 parenright:   right parenthesis            */
+    0x002a,    /* * 0x2A asterisk:    asterisk                     */
+    0x007d,    /* + 0x2B plus:        plus sign                    */
+    0x110c,    /* , 0x2C comma:        comma                        */
+    0x0027,    /* - 0x2D minus:        minus sign                    */
+    0x1103,    /* . 0x2E period:     period                        */
+    0x005b,    /* / 0x2F slash:        slash                        */
+    0x0030,    /* 0x30 0:            0                            */
+    0x0031,    /* 0x31 1:            1                            */
+    0x0032,    /* 0x32 2:            2                            */
+    0x0033,    /* 0x33 3:            3                            */
+    0x0034,    /* 0x34 4:            4                            */
+    0x0035,    /* 0x35 5:            5                            */
+    0x0036,    /* 0x36 6:            6                            */
+    0x0037,    /* 0x37 7:            7                            */
+    0x0038,    /* 0x38 8:            8                            */
+    0x0039,    /* 0x39 9:            9                            */
+    0x110f,    /* : 0x3A colon:        colon                        */
+    0x110f,    /* ; 0x3B semicolon:    semicolon                    */
+    0x110d,    /* < 0x3C less:        less-than sign                */
+    0x005d,    /* = 0x3D equal:        equals sign                    */
+    0x1104,    /* > 0x3E greater:    greater-than sign            */
+    0x007b,    /* ? 0x3F question:    question mark                */
+    0x0040,    /* @ 0x40 at:         commercial at                */
+    0x1106,    /* 0x41 A:            choseong mieum                */
+    0x116e,    /* 0x42 B:            jungseong yu                 */
+    0x1163,    /* 0x43 C:            choseong chieuch             */
+    0x1169,    /* 0x44 D:            choseong ieung                */
+    0x110b,    /* 0x45 E:            choseong ssangtikeut         */
+    0x116d,    /* 0x46 F:            choseong rieul                */
+    0x1167,    /* 0x47 G:            choseong hieuh                */
+    0x1165,    /* 0x48 H:            jungseong o                    */
+    0x1112,    /* 0x49 I:            jungseong ya                 */
+    0x110e,    /* 0x4A J:            jungseong eo                 */
+    0x1111,    /* 0x4B K:            jungseong a                    */
+    0x1168,    /* 0x4C L:            jungseong i                    */
+    0x1173,    /* 0x4D M:            jungseong eu                 */
+    0x1175,    /* 0x4E N:            jungseong u                    */
+    0x1102,    /* 0x4F O:            jungseong yae                */
+    0x1101,    /* 0x50 P:            jungseong ye                 */
+    0x1110,    /* 0x51 Q:            choseong ssangpieup            */
+    0x1164,    /* 0x52 R:            choseong ssangkiyeok         */
+    0x003a,    /* 0x53 S:            choseong nieun                */
+    0x1161,    /* 0x54 T:            choseong ssangsios            */
+    0x1105,    /* 0x55 U:            jungseong yeo                */
+    0x003e,    /* 0x56 V:            choseong phieuph             */
+    0x003c,    /* 0x57 W:            choseong ssangcieuc            */
+    0x1172,    /* 0x58 X:            choseong thieuth             */
+    0x110a,    /* 0x59 Y:            jungseong yo                 */
+    0x003f,    /* 0x5A Z:            choseong khieukh             */
+    0x002d,    /* [ 0x5B bracketleft:  left bracket                 */
+    0x005c,    /* \ 0x5C backslash:    backslash                    */
+    0x003d,    /* ] 0x5D bracketright: right bracket                */
+    0x005e,    /* ^ 0x5E asciicircum:  circumflex accent            */
+    0x0022,    /* _ 0x5F underscore:   underscore                    */
+    0x0060,    /* ` 0x60 quoteleft:    grave accent                 */
+    0x1106,    /* 0x61 a:            choseong  mieum                */
+    0x116e,    /* 0x62 b:            jungseong yu                 */
+    0x1163,    /* 0x63 c:            choseong  chieuch            */
+    0x1169,    /* 0x64 d:            choseong  ieung                */
+    0x110b,    /* 0x65 e:            choseong  tikeut             */
+    0x116d,    /* 0x66 f:            choseong  rieul                */
+    0x1167,    /* 0x67 g:            choseong  hieuh                */
+    0x1165,    /* 0x68 h:            jungseong o                    */
+    0x1112,    /* 0x69 i:            jungseong ya                 */
+    0x110e,    /* 0x6A j:            jungseong eo                 */
+    0x1111,    /* 0x6B k:            jungseong a                    */
+    0x1166,    /* 0x6C l:            jungseong i                    */
+    0x1173,    /* 0x6D m:            jungseong eu                 */
+    0x1175,    /* 0x6E n:            jungseong u                    */
+    0x1102,    /* 0x6F o:            jungseong ae                 */
+    0x1100,    /* 0x70 p:            jungseong e                    */
+    0x1110,    /* 0x71 q:            choseong  pieup                */
+    0x1162,    /* 0x72 r:            choseong  kiyeok             */
+    0x003b,    /* 0x73 s:            choseong  nieun                */
+    0x1161,    /* 0x74 t:            choseong  sios                */
+    0x1105,    /* 0x75 u:            jungseong yeo                */
+    0x002e,    /* 0x76 v:            choseong  phieuph            */
+    0x002c,    /* 0x77 w:            choseong  cieuc                */
+    0x1172,    /* 0x78 x:            choseong  thieuth            */
+    0x1109,    /* 0x79 y:            jungseong yo                 */
+    0x002f,    /* 0x7A z:            choseong  khieukh            */
+    0x005f,    /* { 0x7B braceleft:    left brace                    */
+    0x007c,    /* | 0x7C bar:        vertical bar                 */
+    0x002b,    /* } 0x7D braceright:   right brace                    */
+    0x007e,    /* ~ 0x7E asciitilde:   tilde                        */
+    0x0000    /* 0x7F delete                                     */
+};
+
+static const ucschar hangul_keyboard_table_3finaldvorak[] = {
+    0x0000,    /* 0x00 null                                        */
+    0x0000,    /* 0x01 start of heading                            */
+    0x0000,    /* 0x02 start of text                                */
+    0x0000,    /* 0x03 end of text                                 */
+    0x0000,    /* 0x04 end of transmission                         */
+    0x0000,    /* 0x05 enquiry                                     */
+    0x0000,    /* 0x06 acknowledge                                 */
+    0x0000,    /* 0x07 bell                                        */
+    0x0000,    /* 0x08 backspace                                    */
+    0x0000,    /* 0x09 character tabulation                        */
+    0x0000,    /* 0x0A line feed (lf)                              */
+    0x0000,    /* 0x0B line tabulation                             */
+    0x0000,    /* 0x0C form feed (ff)                              */
+    0x0000,    /* 0x0D carriage return (cr)                        */
+    0x0000,    /* 0x0E shift out                                    */
+    0x0000,    /* 0x0F shift in                                    */
+    0x0000,    /* 0x10 data link escape                            */
+    0x0000,    /* 0x11 device control one                          */
+    0x0000,    /* 0x12 device control two                          */
+    0x0000,    /* 0x13 device control three                        */
+    0x0000,    /* 0x14 device control four                         */
+    0x0000,    /* 0x15 negative acknowledge                        */
+    0x0000,    /* 0x16 synchronous idle                            */
+    0x0000,    /* 0x17 end of transmission block                    */
+    0x0000,    /* 0x18 cancel                                      */
+    0x0000,    /* 0x19 end of medium                                */
+    0x0000,    /* 0x1A substitute                                  */
+    0x0000,    /* 0x1B escape                                      */
+    0x0000,    /* 0x1C information separator four                  */
+    0x0000,    /* 0x1D information separator three                 */
+    0x0000,    /* 0x1E information separator two                    */
+    0x0000,    /* 0x1F information separator one                    */
+    0x0000,    /* 0x20 space                                        */
+    0x11a9,    /* ! 0x21 exclam:        jongseong ssangkiyeok        */
+    0x11c1,    /* " 0x22 quotedbl:      middle dot                  */
+    0x11bd,    /* # 0x23 numbersign:    jognseong cieuc             */
+    0x11b5,    /* $ 0x24 dollar:        jongseong rieul-phieuph     */
+    0x11b4,    /* % 0x25 percent:        jongseong rieul-thieuth     */
+    0x201c,    /* & 0x26 ampersand:     left double quotation mark   */
+    0x11ba,    /* ' 0x27 apostrophe:    choseong  thieuth            */
+    0x0027,    /* ( 0x28 parenleft:     apostrophe                  */
+    0x007e,    /* ) 0x29 parenright:    Tilde                        */
+    0x201d,    /* * 0x2A asterisk:      right double quotation mark  */
+    0x002f,    /* + 0x2B plus:          plus sign                    */
+    0x11af,    /* , 0x2C comma:         comma                        */
+    0x1110,    /* - 0x2D minus:         right parenthesis            */
+    0x1167,    /* . 0x2E period:        period                      */
+    0x0028,    /* / 0x2F slash:         jungseong o                 */
+    0x110f,    /* 0x30 0:             choseong  khieukh            */
+    0x11c2,    /* 0x31 1:             jongseong hieuh             */
+    0x11bb,    /* 0x32 2:             jongseong ssangsios         */
+    0x11b8,    /* 0x33 3:             jongseong pieup             */
+    0x116d,    /* 0x34 4:             jungseong yo                */
+    0x1172,    /* 0x35 5:             jungseong yu                */
+    0x1163,    /* 0x36 6:             jungseong ya                */
+    0x1168,    /* 0x37 7:             jungseong ye                */
+    0x1174,    /* 0x38 8:             jungseong yi                */
+    0x116e,    /* 0x39 9:             jungseong u                 */
+    0x11be,    /* : 0x3A colon:         4                            */
+    0x11b7,    /* ; 0x3B semicolon:     choseong  pieup             */
+    0x11c0,    /* < 0x3C less:          comma                        */
+    0x003c,    /* = 0x3D equal:         greater-than sign            */
+    0x11ac,    /* > 0x3E greater:        period                      */
+    0x0025,    /* ? 0x3F question:      exclamation mark            */
+    0x11b0,    /* @ 0x40 at:            jongseong rieul-kiyeok      */
+    0x11ae,    /* 0x41 A:             jongseong tikeut            */
+    0x002d,    /* 0x42 B:             question mark                */
+    0x0037,    /* 0x43 C:             jongseong khieukh            */
+    0x0030,    /* 0x44 D:             jongseong rieul-pieup        */
+    0x11b2,    /* 0x45 E:             jongseong nieun-cieuc        */
+    0x0035,    /* 0x46 F:             jongseong rieul-mieum        */
+    0x0036,    /* 0x47 G:             jungseong yae                */
+    0x0031,    /* 0x48 H:             0                            */
+    0x1164,    /* 0x49 I:             7                            */
+    0x11bf,    /* 0x4A J:             1                            */
+    0x11aa,    /* 0x4B K:             2                            */
+    0x0039,    /* 0x4C L:             3                            */
+    0x0022,    /* 0x4D M:             double quotation mark        */
+    0x0033,    /* 0x4E N:             minus sign                  */
+    0x11ad,    /* 0x4F O:             8                            */
+    0x11b6,    /* 0x50 P:             9                            */
+    0x11b9,    /* 0x51 Q:             jongseong phieuph            */
+    0x0038,    /* 0x52 R:             jongseong rieul-hieuh        */
+    0x0034,    /* 0x53 S:             jongseong nieun-hieuh        */
+    0x0032,    /* 0x54 T:             jongseong rieul-sios        */
+    0x11b1,    /* 0x55 U:             6                            */
+    0x002e,    /* 0x56 V:             jongseong kiyeok-sios        */
+    0x002c,    /* 0x57 W:             jongseong thikeuth          */
+    0x003f,    /* 0x58 X:             jongseong pieup-sios        */
+    0x11b3,    /* 0x59 Y:             5                            */
+    0x0021,    /* 0x5A Z:             jongseong chieuch            */
+    0x0029,    /* [ 0x5B bracketleft:    left parenthesis            */
+    0x003a,    /* \ 0x5C backslash:     colon                        */
+    0x003e,    /* ] 0x5D bracketright:   less-than sign              */
+    0x003d,    /* ^ 0x5E asciicircum:    equals sign                 */
+    0x00b7,    /* _ 0x5F underscore:    semicolon                    */
+    0x002a,    /* ` 0x60 quoteleft:     asterisk                    */
+    0x11bc,    /* 0x61 a:             jongseong ieung             */
+    0x1109,    /* 0x62 b:             jungseong u                 */
+    0x1106,    /* 0x63 c:             jungseong e                 */
+    0x1102,    /* 0x64 d:             jungseong i                 */
+    0x1175,    /* 0x65 e:             jungseong yeo                */
+    0x1105,    /* 0x66 f:             jungseong a                 */
+    0x1103,    /* 0x67 g:             jungseong eu                */
+    0x110b,    /* 0x68 h:             choseong  nieun             */
+    0x1173,    /* 0x69 i:             choseong  mieum             */
+    0x1166,    /* 0x6A j:             choseong  ieung             */
+    0x1169,    /* 0x6B k:             choseong  kiyeok            */
+    0x1111,    /* 0x6C l:             choseong  cieuc             */
+    0x1112,    /* 0x6D m:             choseong  hieuh             */
+    0x110c,    /* 0x6E n:             choseong  sios              */
+    0x11ab,    /* 0x6F o:             choseong  chieuch            */
+    0x1162,    /* 0x70 p:             choseong  phieuph            */
+    0x11a8,    /* 0x71 q:             jongseong sios              */
+    0x110e,    /* 0x72 r:             jungseong ae                */
+    0x1107,    /* 0x73 s:             jongseong nieun             */
+    0x1100,    /* 0x74 t:             jungseong eo                */
+    0x1161,    /* 0x75 u:             choseong  tikeut            */
+    0x002e,    /* 0x76 v:             jungseong o                 */
+    0x002c,    /* 0x77 w:             jongseong rieul             */
+    0x116e,    /* 0x78 x:             jongseong kiyeok            */
+    0x1165,    /* 0x79 y:             choseong  rieul             */
+    0x1169,    /* 0x7A z:             jongseong mieum             */
+    0x003b,    /* { 0x7B braceleft:     percent sign                */
+    0x005c,    /* | 0x7C bar:            backslash                    */
+    0x002b,    /* } 0x7D braceright:    slash                        */
+    0x203b,    /* ~ 0x7E asciitilde:    reference mark              */
+    0x0000     /* 0x7F delete                                      */
+};
+
+static const ucschar hangul_keyboard_table_3ahnmatae[] = {
+        0x0000    ,              /*    0x00    :    Null            */
+        0x0000    ,              /*    0x01    :    start of heading            */
+        0x0000    ,              /*    0x02    :    start of text            */
+        0x0000    ,              /*    0x03    :    end of text            */
+        0x0000    ,              /*    0x04    :    end of transmission            */
+        0x0000    ,              /*    0x05    :    Enquiry            */
+        0x0000    ,              /*    0x06    :    Acknowledge            */
+        0x0000    ,              /*    0x07    :    Bell            */
+        0x0000    ,              /*    0x08    :    Backspace            */
+        0x0000    ,              /*    0x09    :    character tabulation            */
+        0x0000    ,              /*    0x0A    :    line feed (lf)            */
+        0x0000    ,              /*    0x0B    :    line tabulation            */
+        0x0000    ,              /*    0x0C    :    form feed (ff)            */
+        0x0000    ,              /*    0x0D    :    carriage return (cr)            */
+        0x0000    ,              /*    0x0E    :    shift out            */
+        0x0000    ,              /*    0x0F    :    shift in            */
+        0x0000    ,              /*    0x10    :    data link escape            */
+        0x0000    ,              /*    0x11    :    device control one            */
+        0x0000    ,              /*    0x12    :    device control two            */
+        0x0000    ,              /*    0x13    :    device control three            */
+        0x0000    ,              /*    0x14    :    device control four            */
+        0x0000    ,              /*    0x15    :    negative acknowledge            */
+        0x0000    ,              /*    0x16    :    synchronous idle            */
+        0x0000    ,              /*    0x17    :    end of transmission block            */
+        0x0000    ,              /*    0x18    :    Cancel            */
+        0x0000    ,              /*    0x19    :    end of medium            */
+        0x0000    ,              /*    0x1A    :    Substitute            */
+        0x0000    ,              /*    0x1B    :    Escape            */
+        0x0000    ,              /*    0x1C    :    information separator four            */
+        0x0000    ,              /*    0x1D    :    information separator three            */
+        0x0000    ,              /*    0x1E    :    information separator two            */
+        0x0000    ,              /*    0x1F    :    information separator one            */
+        0x0000    ,              /*    0x20    :    Space            */
+        0x0021    ,          /* ! */    /*    0x21    :    exclam    =    exclamation mark    */
+        0x0022    ,          /* " */    /*    0x22    :    quotedbl    =    quotation mark    */
+        0x0023    ,          /* # */    /*    0x23    :    numbersign    =    number sign    */
+        0x0024    ,          /* $ */    /*    0x24    :    dollar    =    dollar sign    */
+        0x0025    ,          /* % */    /*    0x25    :    percent    =    percent sign    */
+        0x0026    ,          /* & */    /*    0x26    :    ampersand    =    ampersand    */
+        0x002e    ,          /* ' */    /*    0x27    :    apostrophe    =    Period    */
+        0x0028    ,          /* ( */    /*    0x28    :    parenleft    =    left parenthesis    */
+        0x0029    ,          /* ) */    /*    0x29    :    parenright    =    right parenthesis    */
+        0x002a    ,          /* * */    /*    0x2A    :    asterisk    =    asterisk    */
+        0x002b    ,          /*  */    /*    0x2B    :    plus    =    plus sign    */
+        0x11b7    ,          /* , */    /*    0x2C    :    comma    =    jongseong mieum    */
+        0x002d    ,          /* - */    /*    0x2D    :    minus    =    minus sign    */
+        0x11af    ,          /* . */    /*    0x2E    :    period    =    jongseong rieul    */
+        0x11c2    ,          /* / */    /*    0x2F    :    slash    =    jongseong hieuh    */
+        0x0030    ,          /* 0 */    /*    0x30    :    0    =    0    */
+        0x0031    ,          /* 1 */    /*    0x31    :    1    =    1    */
+        0x0032    ,          /* 2 */    /*    0x32    :    2    =    2    */
+        0x0033    ,          /* 3 */    /*    0x33    :    3    =    3    */
+        0x0034    ,          /* 4 */    /*    0x34    :    4    =    4    */
+        0x0035    ,          /* 5 */    /*    0x35    :    5    =    5    */
+        0x0036    ,          /* 6 */    /*    0x36    :    6    =    6    */
+        0x0037    ,          /* 7 */    /*    0x37    :    7    =    7    */
+        0x0038    ,          /* 8 */    /*    0x38    :    8    =    8    */
+        0x0039    ,          /* 9 */    /*    0x39    :    9    =    9    */
+        0x003a    ,          /* : */    /*    0x3A    :    colon    =    colon    */
+        0x116e    ,          /* ; */    /*    0x3B    :    semicolon    =    jungseong u    */
+        0x003c    ,          /* < */    /*    0x3C    :    less    =    less-than sign    */
+        0x003d    ,          /* = */    /*    0x3D    :    equal    =    equals sign    */
+        0x003e    ,          /* > */    /*    0x3E    :    greater    =    greater-than sign    */
+        0x11f9    ,          /* ? */    /*    0x3F    :    question    =    jongseong yet hieuh    */
+        0x0040    ,          /* @ */    /*    0x40    :    at    =    commercial at    */
+        0x1111    ,          /* A */    /*    0x41    :    A    =    choseong phieuph    */
+        0x11f0    ,          /* B */    /*    0x42    :    B    =    jongseong yet ieung    */
+        0x11c1    ,          /* C */    /*    0x43    :    C    =    jongseong phieuph    */
+        0x1110    ,          /* D */    /*    0x44    :    D    =    choseong thieuth    */
+        0x201c    ,          /* E */    /*    0x45    :    E    =    - (quotadouble left)    */
+        0x110f    ,          /* F */    /*    0x46    :    F    =    choseong khiyeok    */
+        0x114c    ,          /* G */    /*    0x47    :    G    =    choseong yet ieung    */
+        0x00b7    ,          /* H */    /*    0x48    :    H    =    middle dot    */
+        0x002f    ,          /* I */    /*    0x49    :    I    =    slash    */
+        0x119e    ,          /* J */    /*    0x4A    :    J    =    jungseong araea    */
+        0x2018    ,          /* K */    /*    0x4B    :    K    =    - (quota left)    */
+        0x2019    ,          /* L */    /*    0x4C    :    L    =    - (quota right)    */
+        0x00b1    ,          /* M */    /*    0x4D    :    M    =    - (plus-minux sign)    */
+        0x11eb    ,          /* N */    /*    0x4E    :    N    =    jongsung ban siot    */
+        0x005b    ,          /* O */    /*    0x4F    :    O    =    [    */
+        0x005d    ,          /* P */    /*    0x50    :    P    =    ]    */
+        0x203b    ,          /* Q */    /*    0x51    :    Q    =    dangujang mark    */
+        0x201d    ,          /* R */    /*    0x52    :    R    =    - (quotadouble right)    */
+        0x110e    ,          /* S */    /*    0x53    :    S    =    choseong chieoch    */
+        0x1159    ,          /* T */    /*    0x54    :    T    =    chosung yet hiheuh    */
+        0x0027    ,          /* U */    /*    0x55    :    U    =    apostrophe    */
+        0x11bf    ,          /* V */    /*    0x56    :    V    =    jongseong khiyeok    */
+        0x1140    ,          /* W */    /*    0x57    :    W    =    chosung ban siot    */
+        0x11c0    ,          /* X */    /*    0x58    :    X    =    jongseong thieuth    */
+        0x003b    ,          /* Y */    /*    0x59    :    Y    =    semicolon    */
+        0x11be    ,          /* Z */    /*    0x5A    :    Z    =    jongseong chieoch    */
+        0x002c    ,          /* [ */    /*    0x5B    :    bracketleft    =    comma    */
+        0x005c    ,          /* \ */    /*    0x5C    :    backslash    =    backslash    */
+        0x003f    ,          /* ] */    /*    0x5D    :    bracketright    =    qustion mark    */
+        0x005e    ,          /* ^ */    /*    0x5E    :    asciicircum    =    circumflex accent    */
+        0x005f    ,          /* _ */    /*    0x5F    :    underscore    =    underscore    */
+        0x0060    ,          /* ` */    /*    0x60    :    quoteleft    =    grave accent    */
+        0x1107    ,          /* a */    /*    0x61    :    a    =    choseong pieup    */
+        0x11bc    ,          /* b */    /*    0x62    :    b    =    jongseong ieung    */
+        0x11b8    ,          /* c */    /*    0x63    :    c    =    jongseong pieup    */
+        0x1103    ,          /* d */    /*    0x64    :    d    =    choseong digeud    */
+        0x1102    ,          /* e */    /*    0x65    :    e    =    choseong nieun    */
+        0x1100    ,          /* f */    /*    0x66    :    f    =    choseong kiyeok    */
+        0x110b    ,          /* g */    /*    0x67    :    g    =    choseong ieung    */
+        0x1165    ,          /* h */    /*    0x68    :    h    =    jungseong eo    */
+        0x1173    ,          /* i */    /*    0x69    :    i    =    jungseong eu    */
+        0x1161    ,          /* j */    /*    0x6A    :    j    =    jungseong a    */
+        0x1175    ,          /* k */    /*    0x6B    :    k    =    jungseong i    */
+        0x1169    ,          /* l */    /*    0x6C    :    l    =    jungseong o    */
+        0x11ab    ,          /* m */    /*    0x6D    :    m    =    jongseong nieun    */
+        0x11ba    ,          /* n */    /*    0x6E    :    n    =    jongseong siot    */
+        0x116d    ,          /* o */    /*    0x6F    :    o    =    jungseong yo    */
+        0x1172    ,          /* p */    /*    0x70    :    p    =    jungseong yu    */
+        0x1106    ,          /* q */    /*    0x71    :    q    =    choseong mieum    */
+        0x1105    ,          /* r */    /*    0x72    :    r    =    choseong rieul    */
+        0x110c    ,          /* s */    /*    0x73    :    s    =    choseong jieut    */
+        0x1112    ,          /* t */    /*    0x74    :    t    =    choseong hieuh    */
+        0x1163    ,          /* u */    /*    0x75    :    u    =    jungseong ya    */
+        0x11a8    ,          /* v */    /*    0x76    :    v    =    jongseong kiyok    */
+        0x1109    ,          /* w */    /*    0x77    :    w    =    choseong siot    */
+        0x11ae    ,          /* x */    /*    0x78    :    x    =    jongseong digeut    */
+        0x1167    ,          /* y */    /*    0x79    :    y    =    jungseong yeo    */
+        0x11bd    ,          /* z */    /*    0x7A    :    z    =    jongseong jieut    */
+        0x007b    ,          /* { */    /*    0x7B    :    braceleft    =    left brace    */
+        0x007c    ,          /* | */    /*    0x7C    :    bar    =    vertical bar    */
+        0x007d    ,          /* } */    /*    0x7D    :    braceright    =    right brace    */
+        0x007e    ,          /* ~ */    /*    0x7E    :    asciitilde    =    Tilde    */
+        0x0000                /*    0x7F    :    delete            */
+};
+
+
+static const HangulCombinationItem hangul_combination_table_3ahnmatae[] = {
+/* need sort
+  { 0x1100110b, 0x1101 }, /* choseong  kiyeok  kiyeok    = ssangkiyeok    **
+  { 0x110b1100, 0x1101 }, /* choseong  kiyeok  kiyeok    = ssangkiyeok    **
+
+  { 0x11031100, 0x1104 }, /* choseong  tikeut  tikeut    = ssangtikeut    **
+  { 0x11001103, 0x1104 }, /* choseong  tikeut  tikeut    = ssangtikeut    **
+
+  { 0x1107110c, 0x1108 }, /* choseong  pieup   pieup    = ssangpieup    **
+  { 0x110c1107, 0x1108 }, /* choseong  pieup   pieup    = ssangpieup    **
+
+  { 0x11091102, 0x110a }, /* choseong  sios    sios    = ssangsios    **
+  { 0x11021109, 0x110a }, /* choseong  sios    sios    = ssangsios    **
+
+  { 0x110c1103, 0x110d }, /* choseong  cieuc   cieuc    = ssangcieuc    **
+  { 0x1103110c, 0x110d }, /* choseong  cieuc   cieuc    = ssangcieuc    **
+//
+  { 0x1112110c, 0x110e }, /* choseong  ch    **
+  { 0x110c1112, 0x110e }, /* choseong    **
+
+  { 0x11121100, 0x110f }, /* choseong  k    **
+  { 0x11001112, 0x110f }, /* choseong    **
+
+  { 0x11121103, 0x1110 }, /* choseong  t    **
+  { 0x11031112, 0x1110 }, /* choseong    **
+
+  { 0x11121107, 0x1111 }, /* choseong  p    **
+  { 0x11071112, 0x1111 }, /* choseong    **
+
+/// 1
+  { 0x11691161, 0x116a }, /* jungseong o       a    = wa        **
+  { 0x11611169, 0x116a }, /* jungseong o       a    = wa        **
+
+  /*` 1169 1161 1175 jungseong o       ae    = wae        **
+
+  { 0x11691175, 0x116c }, /* jungseong o       i    = oe        **
+  { 0x11751169, 0x116c }, /* jungseong o       i    = oe        **
+
+  { 0x116e1165, 0x116f }, /* jungseong u       eo    = weo        **
+  { 0x1165116e, 0x116f }, /* jungseong u       eo    = weo        **
+
+  /*` 116e 1165 1175 jungseong u       e    = we        **
+
+  { 0x116e1175, 0x1171 }, /* jungseong u       i    = wi        **
+  { 0x1175116e, 0x1171 }, /* jungseong u       i    = wi        **
+
+  { 0x11731175, 0x1174 }, /* jungseong eu      i    = yi        **
+  { 0x11751173, 0x1174 }, /* jungseong eu      i    = yi        **
+//
+  { 0x11611175, 0x1162 }, /* jungseong     ae    **
+  { 0x11751161, 0x1162 }, /* jungseong         **
+
+  { 0x11631175, 0x1164 }, /* jungseong     yae    **
+  { 0x11751163, 0x1164 }, /* jungseong         **
+
+  { 0x11651175, 0x1166 }, /* jungseong     e    **
+  { 0x11751165, 0x1166 }, /* jungseong         **
+
+  { 0x11671175, 0x1168 }, /* jungseong     ye    **
+  { 0x11751167, 0x1168 }, /* jungseong         **
+/// jong
+  { 0x11a811bc, 0x11a9 }, /* jongseong kiyeok  kiyeok    = ssangekiyeok    **
+  { 0x11bc11a8, 0x11a9 }, /* jongseong kiyeok  kiyeok    = ssangekiyeok    **
+
+  { 0x11a811ba, 0x11aa }, /* jongseong kiyeok  sios    = kiyeok-sois    **
+  { 0x11ba11a8, 0x11aa }, /* jongseong kiyeok  sios    = kiyeok-sois    **
+
+  { 0x11ab11bd, 0x11ac }, /* jongseong nieun   cieuc    = nieun-cieuc    **
+  { 0x11bd11ab, 0x11ac }, /* jongseong nieun   cieuc    = nieun-cieuc    **
+
+  { 0x11ab11c2, 0x11ad }, /* jongseong nieun   hieuh    = nieun-hieuh    **
+  { 0x11c211ab, 0x11ad }, /* jongseong nieun   hieuh    = nieun-hieuh    **
+
+  { 0x11af11a8, 0x11b0 }, /* jongseong rieul   kiyeok    = rieul-kiyeok    **
+  { 0x11a811af, 0x11b0 }, /* jongseong rieul   kiyeok    = rieul-kiyeok    **
+
+  { 0x11af11ae, 0x11ce }, /*`` r d **
+  { 0x11ae11af, 0x11ce }, /*`` r d **
+
+  { 0x11af11b7, 0x11b1 }, /* jongseong rieul   mieum    = rieul-mieum    **
+  { 0x11b711af, 0x11b1 }, /* jongseong rieul   mieum    = rieul-mieum    **
+
+  { 0x11af11b8, 0x11b2 }, /* jongseong rieul   pieup    = rieul-pieup    **
+  { 0x11b811af, 0x11b2 }, /* jongseong rieul   pieup    = rieul-pieup    **
+
+  { 0x11af11ba, 0x11b3 }, /* jongseong rieul   sios    = rieul-sios    **
+  { 0x11ba11af, 0x11b3 }, /* jongseong rieul   sios    = rieul-sios    **
+
+  { 0x11af11c2, 0x11b6 }, /* jongseong rieul   hieuh    = rieul-hieuh    **
+  { 0x11c211af, 0x11b6 }, /* jongseong rieul   hieuh    = rieul-hieuh    **
+
+  { 0x11b811ba, 0x11b9 }, /* jongseong pieup   sios    = pieup-sios    **
+  { 0x11ba11b8, 0x11b9 }, /* jongseong pieup   sios    = pieup-sios    **
+
+  { 0x11ba11ab, 0x11bb }, /* jongseong sios    sios    = ssangsios    **
+  { 0x11ab11ba, 0x11bb }, /* jongseong sios    sios    = ssangsios    **
+//
+  { 0x11c211bd, 0x11be }, /* jongseong  ch    **
+  { 0x11bd11c2, 0x11be }, /* jongseong    **
+
+  { 0x11c211a8, 0x11bf }, /* jongseong  k    **
+  { 0x11a811c2, 0x11bf }, /* jongseong    **
+
+  { 0x11c211ae, 0x11c0 }, /* jongseong  t    **
+  { 0x11ae11c2, 0x11c0 }, /* jongseong    **
+
+  { 0x11c211b8, 0x11c1 }, /* jongseong  p    **
+  { 0x11b811c2, 0x11c1 }, /* jongseong    **
+
+  { 0x1100110b, 0x1101 }, /* choseong  kiyeok  kiyeok    = ssangkiyeok    **
+  { 0x110b1100, 0x1101 }, /* choseong  kiyeok  kiyeok    = ssangkiyeok    **
+
+  { 0x11031100, 0x1104 }, /* choseong  tikeut  tikeut    = ssangtikeut    **
+  { 0x11001103, 0x1104 }, /* choseong  tikeut  tikeut    = ssangtikeut    **
+
+  { 0x1107110c, 0x1108 }, /* choseong  pieup   pieup    = ssangpieup    **
+  { 0x110c1107, 0x1108 }, /* choseong  pieup   pieup    = ssangpieup    **
+
+  { 0x11091102, 0x110a }, /* choseong  sios    sios    = ssangsios    **
+  { 0x11021109, 0x110a }, /* choseong  sios    sios    = ssangsios    **
+
+  { 0x110c1103, 0x110d }, /* choseong  cieuc   cieuc    = ssangcieuc    **
+  { 0x1103110c, 0x110d }, /* choseong  cieuc   cieuc    = ssangcieuc    **
+
+ /// 3jung
+  { 0x11b611ae, 0x11b4 }, /*`` rh d -> rt **
+  { 0x11ce11c2, 0x11b4 }, /*`` rd h -> rt **
+  { 0x11c011af, 0x11b4 }, /* t r -> rt
+
+  { 0x11b611b8, 0x11b5 }, /*`` rh b -> rp **
+  { 0x11b211c2, 0x11b5 }, /*`` rb h -> rp **
+  { 0x11c111af, 0x11b5 }, /* p r -> rp
+
+  { 0x116a1175, 0x116b},  /* oa ao, i **
+  { 0x116c1161, 0x116b},  /* oi io, a **
+  { 0x11621169, 0x116b},  /* ae ea, o **
+
+  { 0x116f1175, 0x1170},  /* ueo, i **
+  { 0x11711165, 0x1170},  /* ui, eo **
+  { 0x1166116e, 0x1170},  /* e, u **
+
+///// ì°¸ê? ref.
+  ??= o a i
+  ??= u
+  ??= r h d,
+  ??= r h b
+*/
+
+        {    0x11001100    ,    0x110f    }    ,      /*    choseong  kiyeok  kiyeok        khiyeokh    */
+        {    0x11001103    ,    0x1104    }    ,      /*    choseong  kiyeok  tikeut        ssangtikeut    */
+        {    0x1100110b    ,    0x1101    }    ,      /*    choseong  kiyeok  ieung        ssangkiyeok    */
+        {    0x11001112    ,    0x110f    }    ,      /*    choseong  kiyeok  hieuh        khiyeokh    */
+//    {    0x11021102    ,    0x201c    }    ,      /*            - (quotadouble left)    */
+        {    0x11021109    ,    0x110a    }    ,      /*    choseong  nieun    sios        ssangsios    */
+        {    0x11031100    ,    0x1104    }    ,      /*    choseong  tikeut  kiyeok        ssangtikeut    */
+        {    0x11031103    ,    0x1110    }    ,      /*    choseong  tieut   tieut        thieuth    */
+        {    0x1103110c    ,    0x110d    }    ,      /*    choseong  tieut   cieuc        ssangcieuc    */
+        {    0x11031112    ,    0x1110    }    ,      /*    choseong  tieut   hieuh        thieuth    */
+//    {    0x11051105    ,    0x201d    }    ,      /*            - (quotadouble right)    */
+//    {    0x11061106    ,    0x203b    }    ,      /*            dangujang mark    */
+        {    0x11071107    ,    0x1111    }    ,      /*    choseong  pieup   pieup        phieuph    */
+        {    0x1107110c    ,    0x1108    }    ,      /*    choseong  pieup   cieuc        ssangpieup    */
+        {    0x11071112    ,    0x1111    }    ,      /*    choseong  pieup   hieuh        phieuph    */
+        {    0x11091102    ,    0x110a    }    ,      /*    choseong  sios    nieun        ssangsios    */
+//    {    0x11091109    ,    0x1140    }    ,      /*                */
+        {    0x110b1100    ,    0x1101    }    ,      /*    choseong  ieung  kiyeok        ssangkiyeok    */
+        {    0x110c1103    ,    0x110d    }    ,      /*    choseong  cieuc   tikeut        ssangcieuc    */
+        {    0x110c1107    ,    0x1108    }    ,      /*    choseong  cieuc   pieup        ssangpieup    */
+        {    0x110c110c    ,    0x110e    }    ,      /*    choseong  cieuc   cieuc        chieuch    */
+        {    0x110c1112    ,    0x110e    }    ,      /*    choseong  cieuc   hieuh        chieuch    */
+        {    0x11121100    ,    0x110f    }    ,      /*    choseong  hieuh  kiyeok        khiyeokh    */
+        {    0x11121103    ,    0x1110    }    ,      /*    choseong  hieuh   tikeut        thikeuth    */
+        {    0x11121107    ,    0x1111    }    ,      /*    choseong  hieuh   pieup        phieuph    */
+        {    0x1112110c    ,    0x110e    }    ,      /*    choseong  hieuh   cieuc        chieuch    */
+//    {    0x11611161    ,    0x119e    }    ,      /*            arae a    */
+        {    0x11611169    ,    0x116a    }    ,      /*    jungseong a  o        wa    */
+        {    0x11611175    ,    0x1162    }    ,      /*    jungseong  a  i        ae    */
+        {    0x11621169    ,    0x116b    }    ,      /*    ae  o        oae    */
+//    {    0x11631163    ,    0x0027    }    ,      /*            quota    */
+        {    0x11631175    ,    0x1164    }    ,      /*    jungseong ya  i        yae    */
+//    {    0x11651165    ,    0x00b7    }    ,      /*            middle dot    */
+        {    0x1165116e    ,    0x116f    }    ,      /*    jungseong eo  u        weo    */
+        {    0x11651175    ,    0x1166    }    ,      /*    jungseong eo  i        e    */
+        {    0x1166116e    ,    0x1170    }    ,      /*    e  u        wyeo    */
+//    {    0x11671167    ,    0x003b    }    ,      /*            semicolon    */
+        {    0x11671175    ,    0x1168    }    ,      /*    jungseong yeo  i        ye    */
+        {    0x11691161    ,    0x116a    }    ,      /*    jungseong o       a        wa    */
+//    {    0x11691169    ,    0x2019    }    ,      /*            quota right    */
+        {    0x11691175    ,    0x116c    }    ,      /*    jungseong o       i        oe    */
+        {    0x116a1175    ,    0x116b    }    ,      /*    oa  i        oae    */
+        {    0x116c1161    ,    0x116b    }    ,      /*    Oi   a        oae    */
+//    {    0x116c116c    ,    0x005b    }    ,      /*            [    */
+        {    0x116e1165    ,    0x116f    }    ,      /*    jungseong u       eo        weo    */
+//    {    0x116e116e    ,    0x003a    }    ,      /*            colon    */
+        {    0x116e1175    ,    0x1171    }    ,      /*    jungseong u       i        wi    */
+        {    0x116f1175    ,    0x1170    }    ,      /*    Ueo  I        wyeo    */
+        {    0x11711165    ,    0x1170    }    ,      /*    Ui  eo        wyeo    */
+//    {    0x11721172    ,    0x005d    }    ,      /*            ]    */
+//    {    0x11731173    ,    0x002f    }    ,      /*            slash    */
+        {    0x11731175    ,    0x1174    }    ,      /*    jungseong eu      i        yi    */
+        {    0x11751161    ,    0x1162    }    ,      /*    jungseong  I  a        ae    */
+        {    0x11751163    ,    0x1164    }    ,      /*    jungseong  I  ya        yae    */
+        {    0x11751165    ,    0x1166    }    ,      /*    jungseong  I  eo        e    */
+        {    0x11751167    ,    0x1168    }    ,      /*    jungseong  I  yeo        ye    */
+        {    0x11751169    ,    0x116c    }    ,      /*    jungseong o       i        oe    */
+        {    0x1175116e    ,    0x1171    }    ,      /*    jungseong u       i        wi    */
+        {    0x11751173    ,    0x1174    }    ,      /*    jungseong eu      i        yi    */
+//    {    0x11751175    ,    0x2018    }    ,      /*            quota left    */
+        {    0x11a811a8    ,    0x11bf    }    ,      /*    jongseong k   k        khiyeokh    */
+        {    0x11a811af    ,    0x11b0    }    ,      /*    jongseong rieul   kiyeok        rieul-kiyeok    */
+        {    0x11a811ba    ,    0x11aa    }    ,      /*    jongseong kiyeok  sios        kiyeok-sois    */
+        {    0x11a811bc    ,    0x11a9    }    ,      /*    jongseong kiyeok  ieung        ssangekiyeok    */
+        {    0x11a811c2    ,    0x11bf    }    ,      /*    jongseong kiyeok  hieuh            */
+//    {    0x11ab11ab    ,    0x00b1    }    ,      /*            Plus-minus sign    */
+        {    0x11ab11ba    ,    0x11bb    }    ,      /*    jongseong sios    sios        ssangsios    */
+        {    0x11ab11bd    ,    0x11ac    }    ,      /*    jongseong nieun   cieuc        nieun-cieuc    */
+        {    0x11ab11c2    ,    0x11ad    }    ,      /*    jongseong nieun   hieuh        nieun-hieuh    */
+        {    0x11ae11ae    ,    0x11c0    }    ,      /*    jongseong t  t        thieuth    */
+        {    0x11ae11af    ,    0x11ce    }    ,      /*    r d            */
+        {    0x11ae11c2    ,    0x11c0    }    ,      /*    jongseong            */
+        {    0x11af11a8    ,    0x11b0    }    ,      /*    jongseong rieul   kiyeok        rieul-kiyeok    */
+        {    0x11af11ae    ,    0x11ce    }    ,      /*    r d            */
+//    {    0x11af11af    ,    0x003e    }    ,      /*            Great-than    */
+        {    0x11af11b7    ,    0x11b1    }    ,      /*    jongseong rieul   mieum        rieul-mieum    */
+        {    0x11af11b8    ,    0x11b2    }    ,      /*    jongseong rieul   pieup        rieul-pieup    */
+        {    0x11af11ba    ,    0x11b3    }    ,      /*    jongseong rieul   sios        rieul-sios    */
+        {    0x11af11c2    ,    0x11b6    }    ,      /*    jongseong rieul   hieuh        rieul-hieuh    */
+        {    0x11b211c2    ,    0x11b5    }    ,      /*    rb h -> rp            */
+        {    0x11b611ae    ,    0x11b4    }    ,      /*    rh d -> rt            */
+        {    0x11b611b8    ,    0x11b5    }    ,      /*    rh b -> rp            */
+        {    0x11b711af    ,    0x11b1    }    ,      /*    jongseong rieul   mieum        rieul-mieum    */
+//    {    0x11b711b7    ,    0x003c    }    ,      /*            Less-than    */
+        {    0x11b811af    ,    0x11b2    }    ,      /*    jongseong rieul   pieup        rieul-pieup    */
+        {    0x11b811b8    ,    0x11c1    }    ,      /*    jongseong p  p        phieuph    */
+        {    0x11b811ba    ,    0x11b9    }    ,      /*    jongseong pieup   sios        pieup-sios    */
+        {    0x11b811c2    ,    0x11c1    }    ,      /*    jongseong            */
+        {    0x11ba11a8    ,    0x11aa    }    ,      /*    jongseong kiyeok  sios        kiyeok-sois    */
+        {    0x11ba11ab    ,    0x11bb    }    ,      /*    jongseong sios    sios        ssangsios    */
+        {    0x11ba11af    ,    0x11b3    }    ,      /*    jongseong rieul   sios        rieul-sios    */
+        {    0x11ba11b8    ,    0x11b9    }    ,      /*    jongseong pieup   sios        pieup-sios    */
+//    {    0x11ba11ba    ,    0x11eb    }    ,      /*    jongseong        ban sios    */
+        {    0x11bc11a8    ,    0x11a9    }    ,      /*    jongseong kiyeok  kiyeok        ssangekiyeok    */
+//    {    0x11bc11bc    ,    0x11f0    }    ,      /*    jongseong        yet ieung    */
+        {    0x11bd11ab    ,    0x11ac    }    ,      /*    jongseong nieun   cieuc        nieun-cieuc    */
+        {    0x11bd11bd    ,    0x11be    }    ,      /*    jongseong c   c        chieuch    */
+        {    0x11bd11c2    ,    0x11be    }    ,      /*    jongseong            */
+        {    0x11c011af    ,    0x11b4    }    ,      /*    t r -> rt            */
+        {    0x11c111af    ,    0x11b5    }    ,      /*    p r -> rp            */
+        {    0x11c211a8    ,    0x11bf    }    ,      /*    jongseong  k            */
+        {    0x11c211ab    ,    0x11ad    }    ,      /*    jongseong nieun   hieuh        nieun-hieuh    */
+        {    0x11c211ae    ,    0x11c0    }    ,      /*    jongseong  t            */
+        {    0x11c211af    ,    0x11b6    }    ,      /*    jongseong rieul   hieuh        rieul-hieuh    */
+        {    0x11c211b8    ,    0x11c1    }    ,      /*    jongseong  p            */
+        {    0x11c211bd    ,    0x11be    }    ,      /*    jongseong  ch            */
+//    {    0x11c211c2    ,    0x11f9    }    ,      /*    jongseong        yet hieuh    */
+        {    0x11ce11c2    ,    0x11b4    }    ,      /*    rd h -> rt            */
+};
+
 static const ucschar hangul_keyboard_table_2[] = {
     0x0000,     /* 0x00 null                                         */
     0x0000,     /* 0x01 start of heading                             */
--- libhangul-0.1.0.orig/hangul/hangulctype.c
+++ libhangul-0.1.0/hangul/hangulctype.c
@@ -881,6 +881,79 @@ hangul_jongseong_to_choseong(ucschar c)
     return 0;
 }
 
+ucschar
+hangul_shin_choseong_to_shin_jungseong(ucschar c)
+{
+    static ucschar table[] = {
+      0x1100,  /* choseong kiyeok       */
+      0x1101,  /* choseong ssangkiyeok  */
+      0x1102,  /* choseong nieun        */
+      0x1103,  /* choseong tikeut       */
+      0x1104,  /* choseong ssangtikeut      */
+      0x1105,  /* choseong rieul        */
+      0x1174,  /* choseong mieum       ㅁ -> ㅢ  */
+      0x1107,  /* choseong pieup        */
+      0x1108,  /* choseong ssangpieup        */
+      0x1109,  /* choseong sios         */
+      0x110a,  /* choseong ssangsios    */
+      0x110b,  /* choseong ieung        */
+      0x110c,  /* choseong cieuc        */
+      0x110d,  /* choseong ssangcieuc        */
+      0x116e,  /* choseong chieuch     ㅊ -> ㅜ  */
+      0x110f,  /* choseong khieukh      */
+      0x1110,  /* choseong thieuth      */
+      0x1169,  /* choseong phieuph     ㅍ -> ㅗ  */
+      0x1112   /* choseong hieuh        */
+    };
+    if (c < 0x1100 || c > 0x1112)
+        return 0;
+    else if(!(c == 0x1106 || c == 0x110e || c == 0x1111))
+        return 0;
+    else 
+        return table[c - 0x1100];
+}
+
+ucschar
+hangul_shin_jongseong_to_shin_jungseong(ucschar c)
+{
+    static ucschar table[] = {
+      0x116d,  /* jongseong kiyeok        -> jungseong yo  */
+      0x11a9,  /* jongseong ssangkiyeok   ->   */
+      0x11aa,  /* jongseong kiyeok-sios   ->      */
+      0x1168,  /* jongseong nieun         -> jungseong ye        */
+      0x11ac,  /* jongseong nieun-cieuc   ->         */
+      0x11ad,  /* jongseong nieun-hieuh   ->         */
+      0x1175,  /* jongseong tikeut        -> jungseong i        */
+      0x1163,  /* jongseong rieul         ->  jungseong ya        */
+      0x11b0,  /* jongseong rieul-kiyeok  ->        */
+      0x11b1,  /* jongseong rieul-mieum   ->        */
+      0x11b2,  /* jongseong rieul-pieup   ->         */
+      0x11b3,  /* jongseong rieul-sios    ->          */
+      0x11b4,  /* jongseong rieul-thieuth ->       */
+      0x11b5,  /* jongseong rieul-phieuph ->       */
+      0x11b6,  /* jongseong rieul-hieuh   ->         */
+      0x11b7,  /* jongseong mieum         ->         */
+      0x1167,  /* jongseong pieup         -> jungseong yeo      */
+      0x11b9,  /* jongseong pieup-sios    ->          */
+      0x1164,  /* jongseong sios          -> jungseong yae         */
+      0x1161,  /* jongseong ssangsios     -> jungseong a    */
+      0x1172,  /* jongseong ieung         -> jungseong yu        */
+      0x1173,  /* jongseong cieuc         -> jungseong eu        */
+      0x1166,  /* jongseong chieuch       -> jungseong e      */
+      0x116e,  /* jongseong khieukh       -> jungseong u      */
+      0x1162,  /* jongseong thieuth       -> jungseong ae      */
+      0x1165,  /* jongseong phieuph       -> jungseong eo      */
+      0x1169   /* jongseong hieuh         -> jungseong o        */
+    };
+    if (c < 0x11a8 || c > 0x11c2)
+        return 0;
+    else if(c == 0x11a9 || c == 0x11aa || c == 0x11ac || c == 0x11ad || c == 0x11b0 || c == 0x11b1 || c == 0x11b2 || c == 0x11b3 || c == 0x11b4 || c == 0x11b5 || c == 0x11b6 || c == 0x11b7 || c == 0x11b9)
+        return 0;
+    else
+        return table[c - 0x11a8];
+}
+
+
 void
 hangul_jongseong_decompose(ucschar c, ucschar* jong, ucschar* cho)
 {
